### PR TITLE
hw(5): Add basic typechecker and frontend tests

### PR DESCRIPTION
**Pull request by nicdard**

### CODE REVIEW

File: 05/nicdard.ml [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+  ; "wf_typokokbool", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty TBool)
+  ; "wf_typokokreft", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TRef RString)) 
+  ; "wf_typokokreftq", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef RString))
+  ; "array_of_undef_struct", failtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef (RArray (TRef (RStruct "Color")))))
+  ; "array_of_def_struct", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) (Tctxt.add_struct Tctxt.empty "Color" []) (TNullRef (RArray (TRef (RStruct "Color")))))
+  ; "fun_undef_struct", failtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) Tctxt.empty (TNullRef (RFun ([ TRef (RStruct "Color"); TInt; TBool ], RetVoid))))
+  ; "fun_def_struct", passtest @@ (fun () -> Typechecker.typecheck_ty (Ast.no_loc ()) (Tctxt.add_struct Tctxt.empty "Color" []) ((TNullRef (RFun ([ TRef (RStruct "Color"); TInt; TBool ], RetVal TInt)))))
+  ]
+
+let tctxt path (builder : Typechecker.context_builder) =
```
Comment (dbernhard-0x7CD) [inline 05/nicdard.ml:46]: Typechecker.context_builder is not defined in the original typechecker.ml

**Review by dbernhard-0x7CD — CHANGES_REQUESTED** [review]
See comment

File: 05/nicdard.ml (hunk 1 above)
Comment (nicdard) [reply 05/nicdard.ml:49]: You are right sorry! I added a comment with the required type definition to be added to use them, otherwise I can remove them

Comment (nicdard) [reply 05/nicdard.ml:49]: The comment is in the next commit

Comment (dbernhard-0x7CD) [reply 05/nicdard.ml:49]: I'd propose to replace 'Typechecker.context_builder' with the actual type to safe extra editing work. 
But I leave this up to you to decide else LGTM.

File: 05/nicdard.ml [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  ; "struct3.oat", "", "10"
+  ; "struct4.oat", "", "2"
+  ; "struct5.oat", "", "9"
+  ; "global_struct1.oat", "", "10"
+  ; "global_struct2.oat", "", "5"
+  ]
+
+let nicdard_tests = subtype_unit_tests
+  @ typecheck_ty_unit_tests
+  @ context_builder_test Typechecker.create_function_ctxt typecheck_program_correct_ctxt_tests
```
[the remark this answers is not in the document]
Comment (nicdard) [reply 05/nicdard.ml:83]: The idea here is to test that adding function to the ctxt behaves as expected. Actually I have also added builtins into `typecheck_program`.

**Review by nicdard — COMMENTED** [review]
Now it should be fixed, I also added a comment about the function context, let me know if it is fine like this

File: 05/nicdard.ml (hunk 1 above)
Comment (nicdard) [reply 05/nicdard.ml]: Cool thanks!

**Review by dbernhard-0x7CD — APPROVED** [review]

**Pull request merged**